spec/chrome_integration/gtalk.spec.js

```diff
@@ -0,0 +1,237 @@
+ * 4. The refresh_token return from that POST request can be used to get new
+ * freeaccess_tokens
+*/
+var REFRESH_TOKENS = {};
+REFRESH_TOKENS[ALICE_EMAIL] =
+    '1/1BQUvs6QycWFn3CnL-JnmTP6OLl_eDBaIRVf4Kgky4AMEudVrK5jSpoR30zcRFq6';
+REFRESH_TOKENS[BOB_EMAIL] =
+    '1/ZjOcjJskQWyR6mp8L1VpuaR72p4-qVUfJFVfqNlwBc4';
+
+// Wrapper for AJAX request, returns a promise to be fulfilled with the
+// response (expected to be in JSON format).
+function ajax(method, url, data) {
+  return new Promise(function(F, R) {
+    var xhr = new XMLHttpRequest();
+    xhr.open(method, url, true);
+    xhr.setRequestHeader('content-type', 'application/x-www-form-urlencoded');
+    xhr.onload = function() { F(JSON.parse(this.response)); };
+    xhr.send(data);
+  });
+}
+
+var REDIRECT_URL = 'https://www.uproxy.org/oauth-redirect-uri';
+var CLIENT_ID =
+    '746567772449-jkm5q5hjqtpq5m9htg9kn0os8qphra4d.apps.googleusercontent.com';
+
+// Next access token to be returned by OAuthView.launchAuthFlow
+var pendingAccessToken = null;
+
+// Replace default OAuthView with a view that sets the access_token to
+// pendingAccessToken
+var OAuthView = function() {};
+OAuthView.prototype.initiateOAuth = function(redirectURIs, continuation) {
+  continuation({redirect: REDIRECT_URL, state: ''});
+  return true;
+};
+OAuthView.prototype.launchAuthFlow = function(authUrl, stateObj, continuation) {
+  if (!pendingAccessToken) {
+    continuation(undefined, 'No access token found');
+  } else {
+    continuation(REDIRECT_URL + '?access_token=' + pendingAccessToken);
+    pendingAccessToken = null;
+  }
+};
+
+// Returns a Promise that fulfills with an access token.
+function getAccessToken(refreshToken) {
+  return new Promise(function(fulfill, resolve) {
+    var data = 'refresh_token=' + refreshToken +
+        '&client_id=' + CLIENT_ID +
+        '&client_secret=h_hfPI4jvs9fgOgPweSBKnMu' +
+        '&grant_type=refresh_token';
+    ajax('POST', 'https://www.googleapis.com/oauth2/v3/token', data)
+        .then(function(resp) {
+          fulfill(resp.access_token);
+        });
+  });
+}
+
+// Returns a promise that fulfills when socialClient has been logged in
+// using the loginEmail address.  loginEmail must have an associated
+// refresh token.
+function loginAs(socialClient, loginEmail) {
+  var refreshToken = REFRESH_TOKENS[loginEmail.toLowerCase()];
+  if (!refreshToken) {
+    return Promise.reject('No refresh token found for ' + loginEmail);
+  } else if (pendingAccessToken) {
+    return Promise.reject('loginAs cannot be called concurrently.' +
+        '  Please wait until previous login completes');
+  }
+  return getAccessToken(refreshToken).then(function(accessToken) {
+    // Set the accessToken to pendingAccessToken so that it can be used by
+    // our OAuthView.
+    pendingAccessToken = accessToken;
```

Comment:
I agree having the global isn't great, but I'm not sure I understand the proposed solution.  The problem is that there is no way to pass an argument to socialClient.login that will then make its way to OAuthProvider.launchAuthFlow.  If there was then I could pass an email address or just the accessToken and not have any global state.